### PR TITLE
Created an RDDL problem generator

### PR DESCRIPTION
**Pull request by gerardcanal**

### CODE REVIEW

File: rosplan_demos/common/stochastic_domain_turtlebot.pddl [hunk 1]
```diff
@@ -0,0 +1,60 @@
+(define (domain turtlebot)
+
+(:requirements :strips :typing :disjunctive-preconditions); :probabilistic-effects)
```
Comment (m312z) [inline rosplan_demos/common/stochastic_domain_turtlebot.pddl]: Should this requirement be commented out?

File: rosplan_demos/common/deterministic_domain_turtlebot.pddl [hunk 2]
```diff
@@ -0,0 +1,53 @@
+(define (domain turtlebot)
+
```
Comment (m312z) [inline rosplan_demos/common/deterministic_domain_turtlebot.pddl]: It would be good to rename the existing turtlebot domain as well. Perhaps:
temporal_domain_turtlebot.pddl

File: rosplan_demos/common/stochastic_domain_turtlebot.pddl [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+					   0.30 (not (localised ?v))
+		))
+)
+
+;; Localise
+(:action localise
+	:parameters (?v - robot)
+	:precondition (undocked ?v)
+	:effect (probabilistic 0.90 (localised ?v)
+						   0.7 (and))
```
Comment (m312z) [inline rosplan_demos/common/stochastic_domain_turtlebot.pddl]: Does the empty (and) parse fine?

Comment (m312z) [reply rosplan_demos/common/stochastic_domain_turtlebot.pddl]: The way in which failure is done in the dock action looks better.

File: rosplan_knowledge_base/src/VALVisitorOperator.cpp [hunk 4]
```diff
@@ -128,6 +128,7 @@ namespace KCL_rosplan {
 
 	void VALVisitorOperator::visit_effect_lists(VAL::effect_lists * e) {
 
+		eff_time = VAL::E_AT_END;  // Set non-timed add and del effects to be at end in the ROSPlan msgs
```
Comment (m312z) [inline rosplan_knowledge_base/src/VALVisitorOperator.cpp]: I've been looking at this change today, and this is perfect. If we have the same thing for the condition, but "at start", then instant actions are covered. If possible, can you set the default time for *conditions* to be *at start*?

File: rosplan_planning_system/src/PlanDispatch/EsterelPlanDispatcher.cpp [hunk 5]
```diff
@@ -104,6 +104,7 @@ namespace KCL_rosplan {
 		// begin execution
 		finished_execution = false;
 		state_changed = false;
+		bool plan_started = false;
```
Comment (m312z) [inline rosplan_planning_system/src/PlanDispatch/EsterelPlanDispatcher.cpp]: I'm not sure why this change is showing, since the commit is already in #126 (it's fine anyway)

File: rosplan_planning_system/src/ProblemGeneration/ProblemInterface.cpp [hunk 6]
```diff
@@ -19,7 +19,18 @@ namespace KCL_rosplan {
...
-		pddl_problem_generator.knowledge_base = kb;
+
+		// Check problem fle extension to select a problem generator
+		node_handle->param<std::string>("problem_path", problem_path, "common/problem.pddl");
+		std::string extension = (problem_path.size() > 5)? problem_path.substr(problem_path.find_last_of('.')) : "";
+		KCL_rosplan::ProblemGeneratorFactory::ProbGen pg_type;
+		if (extension == ".pddl") pg_type = KCL_rosplan::ProblemGeneratorFactory::PDDL;
+		else if (extension == ".rddl") pg_type = KCL_rosplan::ProblemGeneratorFactory::RDDL;
+		else {
+			ROS_ERROR("KCL: (%s) Unexpected problem file extension %s", ros::this_node::getName().c_str(), extension.c_str());
```
Comment (m312z) [inline rosplan_planning_system/src/ProblemGeneration/ProblemInterface.cpp]: Can this message provide a list of expected types?

File: rosplan_planning_system/src/ProblemGeneration/ProblemInterface.cpp [hunk 7]
```diff
@@ -19,7 +19,18 @@ namespace KCL_rosplan {
 		// connecting to KB
 		std::string kb = "knowledge_base";
 		node_handle->getParam("knowledge_base", kb);
-		pddl_problem_generator.knowledge_base = kb;
+
+		// Check problem fle extension to select a problem generator
+		node_handle->param<std::string>("problem_path", problem_path, "common/problem.pddl");
+		std::string extension = (problem_path.size() > 5)? problem_path.substr(problem_path.find_last_of('.')) : "";
+		KCL_rosplan::ProblemGeneratorFactory::ProbGen pg_type;
```
Comment (m312z) [inline rosplan_planning_system/src/ProblemGeneration/ProblemInterface.cpp]: The problem generation node can also be used to output the problem on a message topic, and not print to file, which is actually the more normal usage to connect to a planner interface. In those instances a user might well leave the problem file parameter blank. The default behaviour according to this code is to use the PDDLGenerator, which I think is reasonable. 

However, an RDDL user might be confused as to how to specify that they want an RDDL problem file. Changing the problem path is an unusual switch.

I think we need to do one of three things:
(1) include an info message, i.e.: "xxx is problem path, so using PDDL problem generation."
(2) include a new parameter to choose the type (PDDL/RDDL)
(3) have two different executables (one for each type)

I would prefer either (2) or (3). I think (2) would mean less refactoring work, while (3) matches more closely the way the planner interface and dispatcher are set up. What do you think?

File: rosplan_knowledge_base/src/VALVisitorOperator.cpp (hunk 4 above)
Comment (m312z) [reply rosplan_knowledge_base/src/VALVisitorOperator.cpp]: Since conditions are parsed before effects for each action, this could be done at the reset of the visitor.

File: rosplan_planning_system/src/ProblemGeneration/ProblemInterface.cpp (hunk 7 above)
Comment (gerardcanal) [reply rosplan_planning_system/src/ProblemGeneration/ProblemInterface.cpp]: I prefer the (2) thinking as a user. It's easier to change an arg or param in the launch file than changing the executable name.

File: rosplan_planning_system/launch/includes/problem_interface.launch [hunk 8]
```diff
@@ -15,6 +16,7 @@
 		<param name="domain_path"    value="$(arg domain_path)" />
 		<param name="problem_path"   value="$(arg problem_path)" />
 		<param name="problem_topic"  value="$(arg problem_topic)" />
+		<param name="planning_language"  value="$(arg planning_language)" />
```
Comment (m312z) [inline rosplan_planning_system/launch/includes/problem_interface.launch]: planning_language vs. planner_language (as it is declared above)?

File: rosplan_planning_system/launch/includes/problem_interface.launch [hunk 9]
```diff
@@ -7,6 +7,7 @@
 	<arg name="domain_path"      default="$(find rosplan_demos)/common/domain_turtlebot_demo.pddl" />
 	<arg name="problem_path"     default="$(find rosplan_demos)/common/problem.pddl" />
 	<arg name="problem_topic"    default="problem_instance" />
+	<arg name="planner_language" default="" /> <!-- Can be RDDL or PDDL -->
```
Comment (m312z) [inline rosplan_planning_system/launch/includes/problem_interface.launch]: Best to set the default here to be the same as the default in the code, for transparency.

File: rosplan_planning_system/launch/includes/problem_interface.launch (hunk 8 above)
Comment (gerardcanal) [reply rosplan_planning_system/launch/includes/problem_interface.launch]: Fixed it this morning
